### ISSUE TITLE
NRE while trying to use datacollectors with vstest

### ISSUE THREAD

**Issue by AM5800** [Microsoft/vstest#849]
## Description
> Describe the issue you've observed

Object reference not set to an instance of an object. While trying to use datacollectors with vstest

## Steps to reproduce
> What steps can reproduce the defect?
> Please share the setup, commandline for vstest.console, sample project, target
> framework etc.

1) Create .Net Core Unit Test Project from template in VS
2) Build
3) In the output directory execute command line:
`dotnet vstest UnitTestProject8.dll /collect:whatever`
It does not matter if 'whatever' is correct collector name or not. Result is always the same

## Expected behavior
> Share the expected output

I would expect output like this:
```
Microsoft (R) Test Execution Command Line Tool Version 15.3.0-preview-20170502-03
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...

Total tests: 1. Passed: 1. Failed: 0. Skipped: 0.
Test Run Successful.
Test execution time: 0,8654 Seconds
```

## Actual behavior
> What is the behavior observed?

Command line output:
```
Microsoft (R) Test Execution Command Line Tool Version 15.3.0-preview-20170502-03
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...
Object reference not set to an instance of an object.
Object reference not set to an instance of an object.
```

Everything runs fine if I remove /collect switch

## Environment
> Please share additional details about the test environment.
> Operating system, Build version of vstest.console

Windows 10 x64
Visual Studio 15.2 (26430.12)
CLI: installed latest version from https://github.com/dotnet/cli namely 2.1.0-preview1-006329

Log files obtained by adding --Diag:log.txt
https://1drv.ms/t/s!AjIeF7rr_c4Ogc5k32dObpi6IbpGHQ

**Comment by harshjain2** [comment]
@AM5800 I ran the same command, here is how it looks like : 
![image](https://user-images.githubusercontent.com/13763032/27083304-f9629d12-4ffc-11e7-86e0-6d2ff38b8029.png)

This is CLI version I've used :
![image](https://user-images.githubusercontent.com/13763032/27083344-1c041576-4ffd-11e7-821a-22f99bd0a512.png)

Can you run the same command by giving the full path to the installation location of latest dotnet CLI.

**Comment by AM5800** [comment]
@harshjain2 when I ran dotnet.exe from vstest\tools it worked!
And the CLI version was exactly the same as yours. 
I also checked my `C:\Program Files\dotnet\sdk` and there is no `2.1.0-preview1-006329` (Although I am sure I installed it). This explains the difference

Anyway, it works. Thank you!

**Comment by AM5800** [comment]
@harshjain2 I am reopening this issue because I think it was a bad idea to close it anyway. I am still experience this issue with .Net Core SDK 2.0.2 which is, I assume, current release version available? And while I can install pre-release versions of SDK - our users might not want it.

To sum up I have 3 questions:
1) Can you reproduce it with 2.0.2? 
2) Is there any workaround available?
3) How does Visual Studio Coverage works then? On my machine it launches the same libraries as I (vstest.console 15.3.0-preview-20170628-02). And I can clearly see VS datacollector in process tree.

I have also created a sample solution for you:
run.cmd - is a command I use to launch tests
console_out.txt - is a console output I receive
diagConsole.txt - is a result of /diag option
[DC202.zip](https://github.com/Microsoft/vstest/files/1467440/DC202.zip)

dotnet --info:

> C:\Work\DC202\TestLib>dotnet --info
> .NET Command Line Tools (2.0.2)
>Product Information:
> Version:            2.0.2
> Commit SHA-1 hash:  a04b4bf512
>Runtime Environment:
> OS Name:     Windows
> OS Version:  10.0.16299
> OS Platform: Windows
> RID:         win10-x64
> Base Path:   C:\Program Files\dotnet\sdk\2.0.2\
> Microsoft .NET Core Shared Framework Host
>  Version  : 2.0.0
>  Build    : e8b8861ac7faf042c87a5c2f9f2d04c98b69f28d

vstest.console version: 15.3.0-preview-20170628-02 (have no idea why it is preview. But the same version I have in C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\Extensions\TestPlatform)

**Comment by smadala** [comment]
@AM5800 Thanks for well made repro 👍 .
1. Yes, I'm able to repro with 2.0.3. Fix will be available in  dotnet 2.1.1. (I think it is going to release with VS 15.5).
2. From logs:
 > TpTrace Verbose: 0 : 4856, 4, 2017/11/13, 16:18:55.477, 262087838229, vstest.console.dll, DotnetDataCollectionLauncher: Full path of dotnet.exe is C:\Program Files\dotnet\dotnet.exe
TpTrace Verbose: 0 : 4856, 4, 2017/11/13, 16:18:55.477, 262087839444, vstest.console.dll, DotnetDataCollectionLauncher: Adding  --runtimeconfig "C:\Program Files\dotnet\sdk\2.0.2\datacollector.runtimeconfig.json" in args
TpTrace Verbose: 0 : 4856, 4, 2017/11/13, 16:18:55.478, 262087840141, vstest.console.dll, DotnetDataCollectionLauncher: Adding  --depsfile "C:\Program Files\dotnet\sdk\2.0.2\datacollector.deps.json" in args
TpTrace Warning: 0 : 4856, 8, 2017/11/13, 16:18:55.501, 262087918950, vstest.console.dll, Data collector standard error line: dotnet exec needs a managed .dll or .exe extension. The application specified was 'C:\Program'
TpTrace Error: 0 : 4856, 6, 2017/11/13, 16:18:55.504, 262087927297, vstest.console.dll, Data collector exited with error: 'dotnet exec needs a managed .dll or .exe extension. The application specified was 'C:\Program'
' 

Looks like issue is dupe of https://github.com/Microsoft/vstest/issues/902.
**Workaround:** Don't have space in path to `dotnet` folder.
3. VS Code coverage uses datacollector.exe("C:\Program Files (x86)\Microsoft Visual Studio\Enterprise\Common7\IDE\Extensions\TestPlatform\datacollector.exe"), Where as `dotnet vstest --collect:"SimpleDC"` use datacollector.dll ("C:\Program Files\dotnet\sdk\2.0.3\datacollector.dll").
Preview version is known issue https://github.com/Microsoft/vstest/issues/1134.
